File: Lab1/CheckSum.java
/**
 * CheckSum.java: Implementación del algoritmo de Checksum
 * 
 * @version 1.0.0
 * @autor Javier Valle
 * @author Javier Mombiela
 * @date 01-08-2023
 * 
 */


 /**
  * CheckSumSender: Clase que implementa el algoritmo de Checksum para el envio de mensajes
  */
class CheckSumSender {

    private int blockSize = 8;
    public String message = "";
    public String dividedMessage = "";
    public String checkSum = "";
    public String messageWithCheckSum = "";


    /**
     * Constructor de la clase CheckSumSender
     * @param message Mensaje a enviar
     */
    public CheckSumSender(String message) {
        this.message = message;
    }

    /**
     * Divide el mensaje en bloques de 8 bits y agrega padding si es necesario
     * @return El mensaje dividido en bloques de 8 bits
     */
    public String divideMessage() {
        int messageLength = message.length();
        int paddingNeeded = blockSize - (messageLength % blockSize);
        
        // Only add padding if needed
        if (paddingNeeded < blockSize) {
            StringBuilder paddedMessage = new StringBuilder(message);
            for (int i = 0; i < paddingNeeded; i++) {
                paddedMessage.append("0");
            }
            message = paddedMessage.toString();
        }

        // Divide the message into blocks of blockSize
        StringBuilder dividedMessage = new StringBuilder();
        for (int i = 0; i < message.length(); i += blockSize) {
            int endIndex = Math.min(i + blockSize, message.length());
            String block = message.substring(i, endIndex);
            dividedMessage.append(block).append(" ");
        }

        return dividedMessage.toString().trim();
    }

    /**
     * Calcula el CheckSum del mensaje
     * @return El CheckSum del mensaje
     */
    public String calculateCheckSum() {
        String[] blocks = this.dividedMessage.split(" ");

        // inicializa la suma en 0
        StringBuilder sumBinary = new StringBuilder("00000000");

        for (String block : blocks) {
            int blockValue = Integer.parseInt(block, 2); // Convert the block to an integer
            int sumValue = Integer.parseInt(sumBinary.toString(), 2); // Convert the sum to an integer

            int tempSum = blockValue + sumValue;

            // revisa si hay bits flotantes y los agrega al bit menos significativo
            if (tempSum > 255) {
                tempSum = (tempSum & 255) + 1; 
            }

            // Convierte la nueva suma a binario y agrega padding si es necesario
            String newSumBinary = Integer.toBinaryString(tempSum);
            while (newSumBinary.length() < 8) {
                newSumBinary = "0" + newSumBinary;
            }

            sumBinary = new StringBuilder(newSumBinary);
        }

        // System.out.println("Sum Binary: " + sumBinary);
        // Toma el complemento a 1 de la suma final
        for (int i = 0; i < sumBinary.length(); i++) {
            char bit = sumBinary.charAt(i);
            sumBinary.setCharAt(i, bit == '0' ? '1' : '0');
        }

        return sumBinary.toString();
    }

    /**
     * Imprime el mensaje, el mensaje dividido, el CheckSum y el mensaje final
     */
    public void sendMessage() {
        this.dividedMessage = this.divideMessage();
        System.out.println("Mensaje          -> " + this.message);
        System.out.println("Mensaje Dividido -> " + this.dividedMessage);
        this.checkSum = this.calculateCheckSum();
        System.out.println("CheckSum         -> " + this.checkSum);
        this.messageWithCheckSum = this.dividedMessage + " " + this.checkSum;
        System.out.println("Mensaje final    -> " + this.messageWithCheckSum);
    }
}


/**
 * CheckSumReceiver: Clase que implementa el algoritmo de Checksum para la recepción de mensajes
 */
class CheckSumReceiver {

    public String messageReceived = "";
    public String chekedSum = "";
    public CheckSumSender sender;


    /**
     * Constructor de la clase CheckSumReceiver
     * @param message Mensaje recibido
     */
    public CheckSumReceiver(String message) {
        this.messageReceived = message;
    }

    /**
     * Calcula el CheckSum del mensaje recibido
     * @return El CheckSum del mensaje recibido
     */
    public String checkTheSum() {
       String[] blocks = this.messageReceived.split(" ");

        // Inicializa la suma a 0
        StringBuilder sumBinary = new StringBuilder("00000000");

        for (String block : blocks) {
            int blockValue = Integer.parseInt(block, 2); // Convert the block to an integer
            int sumValue = Integer.parseInt(sumBinary.toString(), 2); // Convert the sum to an integer

            int tempSum = blockValue + sumValue;

            // Revisa si hay bits flotantes y los agrega al bit menos significativo
            if (tempSum > 255) {
                tempSum = (tempSum & 255) + 1; 
            }

            // Convierte la nueva suma a binario y agrega padding si es necesario
            String newSumBinary = Integer.toBinaryString(tempSum);
            while (newSumBinary.length() < 8) {
                newSumBinary = "0" + newSumBinary;
            }

            sumBinary = new StringBuilder(newSumBinary);
        }

        // System.out.println("Sum Binary: " + sumBinary);
        // Toma el complemento a 1 de la suma final
        for (int i = 0; i < sumBinary.length(); i++) {
            char bit = sumBinary.charAt(i);
            sumBinary.setCharAt(i, bit == '0' ? '1' : '0');
        }

        return sumBinary.toString();
    }

    /**
     * Analiza el CheckSum del mensaje recibido
     */
    public void analyzeSum() {
        if (this.chekedSum.equals("00000000")) {
            System.out.println("Resultado         -> La trama no contiene errores!");
        } else {
            System.out.println("Resultado         -> La trama contiene errores y sera descartada!");
        }
    }

    /**
     * Imprime el mensaje recibido, el CheckSum recibido y el resultado del análisis
     */
    public void receiveMessage() {
        System.out.println("Mensaje recibido  -> " + this.messageReceived);
        this.chekedSum = this.checkTheSum();
        System.out.println("CheckSum recibido -> " + this.chekedSum);
        this.analyzeSum();
    }
}


/**
 * CheckSum: Clase principal que ejecuta el programa
 */
class CheckSum {
    public static void main(String[] args) {
        System.out.println("\n----ENVIANDO MENSAJE----");
        CheckSumSender sender = new CheckSumSender("10101010");
        // CheckSumSender sender = new CheckSumSender("1010100100111001");
        // CheckSumSender sender = new CheckSumSender("10000100001001001110001010011001");
        sender.sendMessage();
        String checksum = sender.checkSum;

        System.out.println("\n----RECIBIENDO MENSAJE----");
        CheckSumReceiver receiver = new CheckSumReceiver("10101011 " + checksum);
        receiver.receiveMessage();
    }
}

File: Lab1/CheckSum.py
"""
CheckSum.py: Implementacion del algoritmo de Checksum para la deteccion de errores en tramas de datos.

Autores: Javier Valle, Javier Mombiela
Fecha: 01/08/2023
"""


"""
CheckSumSender: Clase que implementa el algoritmo de Checksum para la deteccion de errores en tramas de datos.
"""
class CheckSumSender:

    def __init__(self, message):
        """
        Constructor de la clase CheckSumSender.

        :param message: Trama de datos a enviar.
        """

        self.blockSize = 8
        self.message = message
        self.dividedMessage = ""
        self.checkSum = ""
        self.messageWithCheckSum = ""

    def divideMessage(self):
        """
        metodo para dividir la trama de datos en bloques de 8 bits.

        :return: Trama de datos dividida en bloques de 8 bits.
        """

        messageLength = len(self.message)
        paddingNeeded = self.blockSize - (messageLength % self.blockSize)

        # Only add padding if needed
        if paddingNeeded < self.blockSize:
            paddedMessage = self.message + "0" * paddingNeeded
            self.message = paddedMessage

        # Divide the message into blocks of blockSize
        dividedMessage = " ".join([self.message[i:i+self.blockSize] for i in range(0, len(self.message), self.blockSize)])
        self.dividedMessage = dividedMessage.strip()
        return self.dividedMessage

    def calculateCheckSum(self):
        """
        Metodo para calcular el Checksum de la trama de datos.

        :return: Checksum de la trama de datos.
        """

        blocks = self.dividedMessage.split(" ")

        # initialize the sum to 0
        sumBinary = "00000000"

        for block in blocks:
            blockValue = int(block, 2)  # Convert the block to an integer
            sumValue = int(sumBinary, 2)  # Convert the sum to an integer

            tempSum = blockValue + sumValue

            # check for carry and add it to the least significant bit
            if tempSum > 255:
                tempSum = (tempSum & 255) + 1

            # Convert the new sum to binary and add padding if necessary
            newSumBinary = bin(tempSum)[2:].zfill(8)

            sumBinary = newSumBinary

        # Take the one's complement of the final sum
        sumBinary = "".join('1' if bit == '0' else '0' for bit in sumBinary)

        self.checkSum = sumBinary
        return self.checkSum

    def sendMessage(self):
        """
        Metodo para enviar la trama de datos con el Checksum.
        """

        self.dividedMessage = self.divideMessage()
        print("Mensaje          ->", self.message)
        print("Mensaje Dividido ->", self.dividedMessage)
        self.checkSum = self.calculateCheckSum()
        print("CheckSum         ->", self.checkSum)
        self.messageWithCheckSum = self.dividedMessage + " " + self.checkSum
        print("Mensaje final    ->", self.messageWithCheckSum)

"""
CheckSumReceiver: Clase que implementa el algoritmo de Checksum para la deteccion de errores en tramas de datos.
"""
class CheckSumReceiver:

    def __init__(self, message):
        """
        Constructor de la clase CheckSumReceiver.

        :param message: Trama de datos a recibir.
        """

        self.messageReceived = message
        self.chekedSum = ""
        self.sender = None

    def checkTheSum(self):
        """
        Metodo para calcular el Checksum de la trama de datos recibida.

        :return: Checksum de la trama de datos recibida.
        """

        blocks = self.messageReceived.split(" ")

        # initialize the sum to 0
        sumBinary = "00000000"

        for block in blocks:
            blockValue = int(block, 2)  # Convert the block to an integer
            sumValue = int(sumBinary, 2)  # Convert the sum to an integer

            tempSum = blockValue + sumValue

            # check for carry and add it to the least significant bit
            if tempSum > 255:
                tempSum = (tempSum & 255) + 1

            # Convert the new sum to binary and add padding if necessary
            newSumBinary = bin(tempSum)[2:].zfill(8)

            sumBinary = newSumBinary

        # Take the one's complement of the final sum
        sumBinary = "".join('1' if bit == '0' else '0' for bit in sumBinary)

        self.chekedSum = sumBinary
        return self.chekedSum

    def analyzeSum(self):
        """
        Metodo para analizar el Checksum de la trama de datos recibida.
        """

        if self.chekedSum == "00000000":
            print("Resultado         -> La trama no contiene errores!")
        else:
            print("Resultado         -> La trama contiene errores y será descartada!")

    def receiveMessage(self):
        """
        Metodo para recibir la trama de datos con el Checksum.
        """

        print("Mensaje recibido  ->", self.messageReceived)
        self.chekedSum = self.checkTheSum()
        print("CheckSum recibido ->", self.chekedSum)
        self.analyzeSum()

"""
CheckSum: Clase que implementa el algoritmo de Checksum para la deteccion de errores en tramas de datos.
"""
class CheckSum:

    def __init__(self):
        """
        Constructor de la clase CheckSum.
        """

        print("----ENVIANDO MENSAJE----")
        message = "1010100100111001"
        sender = CheckSumSender(message)
        sender.sendMessage()
        checksum = sender.checkSum

        print("\n----RECIBIENDO MENSAJE----")
        message2 = "10101001 00111001"
        receiver = CheckSumReceiver(message2 + " " + checksum)
        receiver.receiveMessage()

if __name__ == '__main__':
    CheckSum()
